Fix Azure OpenAI API Version Precedence and Structured Output Handling

colesmcintosh opened the pull request:
## Title  

Fix Azure OpenAI API Version Precedence and Structured Output Handling  

## Relevant Issues  

- Fix issue where Azure OpenAI API version specified in LiteLLM proxy config was overridden by client-provided API version  
- Fix issue with structured output and function calling in Azure OpenAI models  

## Pre-Submission Checklist  

- [x] Added unit tests in the `tests/litellm/` directory to validate API version precedence  
- [x] Verified tests are passing locally
<img width="666" alt="Screenshot 2025-03-11 at 9 16 51 PM" src="https://github.com/user-attachments/assets/001c4222-6fa5-4d9f-9fbe-7211667a5aca" />
<img width="667" alt="Screenshot 2025-03-11 at 9 17 17 PM" src="https://github.com/user-attachments/assets/cfa5d297-e1ee-42cf-b24b-97f16c35a578" />
<img width="672" alt="Screenshot 2025-03-11 at 9 17 51 PM" src="https://github.com/user-attachments/assets/cd85a504-02b7-4722-8fce-fd1f441ffe20" />



- [x] PR scope is isolated to fixing API version handling and structured output issues  

## Type  

🐛 Bug Fix  
✅ Test  

## Changes  

### 1. API Version Handling in Pre-Call Utils  
- Ensured API version specified in LiteLLM proxy configuration takes precedence over client-provided versions  
- Improved API version configuration flexibility by prioritizing proxy settings over query parameters  
- Added unit tests to verify API version precedence  

### 2. Fix Azure OpenAI Structured Output Issue  
- Adjusted model configuration to disable `supports_parallel_function_calling` for Azure models that do not support it  
- Improved error messaging to guide users toward using the correct API version and parameters  

### 3. Testing  
- Added unit test to verify API version precedence in pre-call utilities  
- Confirmed fix by reproducing the issue and validating against test cases  

## Additional Context  

This PR addresses two high-priority issues for a LiteLLM user leveraging Azure OpenAI models. The customer encountered errors when using `tools` and `response_format` parameters, which were not supported due to incorrect API version handling. The fix ensures that users have a consistent experience when integrating with Azure OpenAI via LiteLLM.

colesmcintosh committed:
test(azure): add structured output and tool calling tests for Azure OpenAI

colesmcintosh committed:
fix(proxy): improve api_version handling in pre-call utils

Enhance API version configuration by prioritizing proxy config settings over query parameters. This allows more flexible and precise API version specification for individual models.

colesmcintosh committed:
fix(azure): update parallel function calling support for Azure models

Modify model configuration to set supports_parallel_function_calling to false for multiple Azure OpenAI models, reflecting their current capabilities.

https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/function-calling

colesmcintosh committed:
feat(proxy): enhance API version configuration in pre-call utils

Add support for setting API version from model configuration, ensuring it takes precedence over query parameters when available. This improves the flexibility of API version specification in the proxy.

colesmcintosh committed:
test(proxy): add test for API version precedence in pre-call utils

Implement a unit test to verify that API versions specified in proxy configuration take precedence over client-provided query parameters, ensuring consistent and configurable API version handling.

ishaan-jaff commented on litellm/proxy/litellm_pre_call_utils.py:
```diff
@@ -464,7 +464,22 @@ async def add_litellm_data_to_request(  # noqa: PLR0915
 
     from litellm.proxy.proxy_server import llm_router, premium_user
 
-    safe_add_api_version_from_query_params(data, request)
```
add this logic within `safe_add_api_version_from_query_params` 

- first check if the user dynamically passed an api version 
- then read from llm_router (not from proxy_config)

ishaan-jaff commented on tests/test_azure_structured_output.py:
```diff
@@ -0,0 +1,335 @@
+        api_version="2024-12-01-preview"
+    )
+
+def test_basic_structured_output(azure_client):
```
confused, what is this testing ? is it testing against the litellm proxy

colesmcintosh replied:
The goal is to test the actual API and catch any potential breaking changes, like name change from "function" to "tool". That way, if something stops working, we can quickly tell whether it’s an issue with the package or a change Azure made on their end.

ishaan-jaff reviewed (changes requested):
reviewed

colesmcintosh committed:
feat(proxy): enhance safe_add_api_version_from_query_params to prioritize proxy config

Updated the safe_add_api_version_from_query_params function to include an optional proxy_config parameter, allowing API version to be set from model configuration

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=9160) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=9160) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=9160) it.</sub>
